webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/servers/RemoveNodeService.scala

```diff
@@ -327,69 +493,192 @@ class RemoveNodeServiceImpl(
+                        for {
+                          nodeGroup    <- roNodeGroupRepository.getNodeGroup(nodeGroupId).map(_._1)
+                          updatedGroup =  nodeGroup.copy(serverList = nodeGroup.serverList - nodeId)
+                          msg          =  Some("Automatic update of group due to deletion of node " + nodeId.value)
+                          diff         <- (if(nodeGroup.isSystem) {
+                                            woNodeGroupRepository.updateSystemGroup(updatedGroup, modId, actor, msg)
+                                          } else {
+                                            woNodeGroupRepository.update(updatedGroup, modId, actor, msg)
+                                          }).chainError(s"Could not update group '${nodeGroup.id.value}' to remove node '${nodeId.value}'")
+                        } yield {
+                          diff
+                        }
+                      }
     } yield {
-      res
+      deleted.flatten
     }
   }
+}
 
-  /**
+
+/*
+ * Close expected reports for node.
+ * Also delete nodes_info for that node.
+ */
+class CloseNodeConfiguration(expectedReportsRepository: UpdateExpectedReportsRepository) extends PostNodeDeleteAction {
+  override def run(nodeId: NodeId, mode: DeleteMode, info: Option[NodeInfo], status: Set[InventoryStatus]): UIO[Unit] = {
+    for {
+      _ <- NodeLoggerPure.Delete.debug(s"  - close expected reports for '${nodeId.value}'")
+      _ <- expectedReportsRepository.closeNodeConfigurationsPure(nodeId).catchAll(err =>
+             NodeLoggerPure.Delete.error(s"Error when closing expected reports for node ${(nodeId, info).name}")
+           ).unit
+      _ <- expectedReportsRepository.deleteNodeInfos(nodeId).catchAll(err => NodeLoggerPure.Delete.error(err.msg))
+    } yield ()
+  }
+}
+// when the node is a policy server, delete directive/rule/group related to it
+class DeletePolicyServerPolicies(policyServerManagement: PolicyServerManagementService) extends PostNodeDeleteAction {
+  override def run(nodeId: NodeId, mode: DeleteMode, info: Option[NodeInfo], status: Set[InventoryStatus]): UIO[Unit] = {
+    // we can avoid to do LDAP requests if we are sure the node wasn't a policy server
+    info.map(_.isPolicyServer) match {
+      case Some(false) =>
+        UIO.unit
+      case _           =>
+        NodeLoggerPure.Delete.debug(s"  - delete relay related policies in LDAP'${nodeId.value}'") *>
+        policyServerManagement.deleteRelaySystemObjectsPure(nodeId).catchAll(err =>
+          NodeLoggerPure.Delete.error(s"Error when deleting system objects (groups, directives, rules) related to relay server for node ${(nodeId, info).name}: ${err.fullMsg}")
+        )
+    }
+  }
+}
+
+// clean up certification key status (only in move mode, not erase)
+class KeyStatus(ldap: LDAPConnectionProvider[RwLDAPConnection], deletedDit: InventoryDit) extends PostNodeDeleteAction {
```

Comment:
this name is *really* misleading. I thought it was a case class when first encountering it.
KeyStatusDeletion?